### PR TITLE
Feature 832 custom rust target

### PR DESCRIPTION
**Pull request by tmfink**
Addresses #832.

Instead of specifying whether or not to use stable, specify the Rust
release to support (one of several stable/beta releases or nightly).
The `--unstable-rust` option is still accepted and implies `--rust-target nightly`.

The definitions of `RustTarget` and `RustFeatures` are created with
macros.

In order to keep the test outputs the same,
`bindgen-flags: --rust-target 1.0` was added to test headers.

**Todo:**

- [x] Create `RustFeatures`/`RustTarget` structs
- [x] Replace uses of `unstable` with `RustFeatures` query
- [x] Add new tests
- [x] Fix doc comments TODOs

### CODE REVIEW

**Comment by fitzgen** [comment]
One more thing: can you [rebase and squash these commits into a single commit](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) on top of master? Thanks!

File: src/features.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+#![deny(missing_docs)]
+#![deny(warnings)]
+#![deny(unused_extern_crates)]
+
+use std::io;
+
+/// Define RustTarget struct definition, Default impl, and conversions
+/// between RustTarget and String.
+macro_rules! rust_target_def {
+    ( $( $release:ident => $value:expr => $attrs:meta; )* ) => {
```
Comment (tmfink) [inline src/features.rs]: Wow, I didn't realize you could do that.

File: src/features.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        impl RustTarget {
+            #[allow(dead_code)]
+            /// Create a `RustTarget` from a string.
+            ///
+            /// * The stable/beta versions of Rust are of the form "1.0",
+            /// "1.19", etc.
+            /// * The nightly version should be specified with "nightly".
+            pub fn from_str<'a>(s: &'a str) -> io::Result<RustTarget> {
```
Comment (tmfink) [inline src/features.rs]: Nice, I did not know that existed.

File: src/options.rs [hunk 3]
```diff
@@ -161,6 +161,10 @@ pub fn builder_from_flags<I>
                 .takes_value(true)
                 .multiple(true)
                 .number_of_values(1),
+            Arg::with_name("rust-target")
+                .long("rust-target")
+                .help("Version of the Rust compiler to target.")
```
Comment (tmfink) [inline src/options.rs]: I decided to use the macros.

**Comment by tmfink** [comment]
Looks like there were some issues with my rebase--I'll fix it tomorrow.

**Comment by fitzgen** [comment]
> Looks like there were some issues with my rebase--I'll fix it tomorrow.

Ok! Let me know if you have any questions or need any help or anythign :)

**Comment by tmfink** [comment]
I squashed everything, but the downside is that it lost "git copy" trick that I used to keep the provenance of the "1.0" tests. Now Git thinks that I made the headers/expectations from scratch.

**Comment by fitzgen** [comment]
While that's too bad, ultimately, I don't think it's a huge deal.

File: src/lib.rs [hunk 4]
```diff
@@ -80,6 +82,8 @@ mod codegen {
     include!(concat!(env!("OUT_DIR"), "/codegen.rs"));
 }
 
+pub use features::{RustTarget, LATEST_STABLE_RUST, RUST_TARGET_STRINGS};
```
Comment (fitzgen) [inline src/lib.rs]: I don't see `RUST_TARGET_STRINGS` used in this module, is it a leftover from some earlier iteration?

Comment (fitzgen) [inline src/lib.rs]: Huh, well there are no warnings in the travis ci compilation (and I guess we deny(warnings) anyhow).

**Comment by fitzgen** [comment]
@bors-servo r+

Thanks @tmfink !

**Pull request merged**